File: CodeForces/Hello 2020/B.py
N = int(input())
seqs = []
for i in range(N):
    vals = list(map(int, input().split()))[1:]
    seqs.append(vals)

works = 0
los = []
his = []

ans = 0

for seq in seqs:
    if sorted(seq, reverse=True) != seq: works += 1
    else:
        lo = min(seq)
        hi = max(seq)

        los.append(lo)
        his.append(hi)

los = sorted(los)
his = sorted(his)

for lo, hi in zip(los, his):
    l = 0
    h = len(his)-1

    
    while l < h:
        m = (l + h) // 2
        # print(l, m, h)
        if his[m] > lo:
            h = m
        else:
            l = m + 1

    

    if his[l] > lo:
        print("AS LOW", lo, l, len(los) - l)
        ans += (len(los) - l)

    l = 0
    h = len(los)-1
    while l < h:
        m = (l + h + 1) // 2
        if los[m] < hi:
            l = m
        else:
            h = m - 1
    if los[l] < hi:
        print("AS HI", hi, l + 1)
        ans += l + 1
ans //= 2
print(ans)
ans += 2 * works * N
print(ans)

File: CodeForces/Hello 2020/B.cpp
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
#include <bits/stdc++.h>
 
using namespace __gnu_pbds;
using namespace std;
 
typedef
tree<
  pair<int,int>,
  null_type,
  less<pair<int,int>>,
  rb_tree_tag,
  tree_order_statistics_node_update>
ordered_set;

#define FOR(i,a,b) for (int i = (a); i < (b); ++i)
#define F0R(i,a) FOR(i,0,a)
#define ROF(i,a,b) for (int i = (b)-1; i >= (a); --i)
#define R0F(i,a) ROF(i,0,a)

using ll = long long;
ll INF = LLONG_MAX;
 
using vc = vector<char>;
using vi = vector<int>;

int main() {
    ios::sync_with_stdio(0), cin.tie(0), cout.tie(0);
    int N; cin >> N;
    ordered_set los;
    ordered_set his;

    ll gay = 0;
    ll ans = 0;
    F0R(j, N) {
        int M; cin >> M;
        vi arr (M);
        F0R(i, M) cin >> arr[i];
        reverse(arr.begin(), arr.end());
        if (!is_sorted(arr.begin(), arr.end())) {
            ++gay;
        } else {
            int lo = arr[0];
            int hi = arr[0];
            F0R(i, M) {
                lo = min(lo, arr[i]);
                hi = max(hi, arr[i]);
            }

            

            // as LOW
            ans += his.size() - his.order_of_key({lo, his.size()+1});

            // as HI
            ans += los.order_of_key({hi, -1});
            if (lo < hi) ++ans;
            los.insert({lo, los.size()});
            his.insert({hi, his.size()});
        }
    }
    //cout << ans << " " << gay << endl;
    ans += gay * gay;
    ans += gay * (N - gay);
    ans += (N-gay) * gay;
    cout << ans  << endl;
}